File: apps/Camera-Vitals/iOS/CameraVitals/Pipeline/RPPGPipeline.swift
import CoreGraphics
import CoreVideo
import Foundation

/// Buffer/ROI sizes reported alongside results, so overlays can map ROI → preview.
struct FaceState {
    let faceFound: Bool
    let faceBox: CGRect?          // pixel coords in the (portrait) buffer
    let bufferSize: CGSize
    let lowLight: Bool
    let framesFilled: Int
    let capacity: Int
}

struct VitalsUpdate {
    let bpm: Double?
    let quality: Double
    let waveform: [Float]
    let latencyMs: Double
    let warmupProgress: Double   // 0...1 — stitch buffer filling toward the first reading
}

protocol RPPGPipelineDelegate: AnyObject {
    func pipeline(_ pipeline: RPPGPipeline, didUpdateFace state: FaceState)
    func pipeline(_ pipeline: RPPGPipeline, didUpdateVitals update: VitalsUpdate)
}

/// Orchestrates the temporal pipeline: face crop → ring buffer → sliding-window inference → HR.
/// `process(_:)` is called on the capture queue; inference runs on its own queue with a
/// single-flight `busy` gate (drop windows under back-pressure, never queue them).
final class RPPGPipeline {
    weak var delegate: RPPGPipelineDelegate?

    private let model: RPPGModel
    private let tracker = FaceROITracker()
    private let ring = FrameRingBuffer(capacity: AppConfig.framesIn, frameLen: AppConfig.frameFloatCount)
    private let inferenceQueue = DispatchQueue(label: "rppg.inference", qos: .userInitiated)

    private var frameCounter = 0
    private var detectCounter = 0
    private var lastBox: CGRect?
    private var busy = false
    private let busyLock = NSLock()

    // Stitched rPPG samples + display waveform + BPM smoothing — touched only on the inference queue.
    private var rawWaveform = [Float]()       // raw model outputs, stitched (rolling)
    private var displayWaveform = [Float]()   // band-passed, for the chart
    private var bpmFilter = MedianEMA(size: 5, alpha: 0.3)
    private var inferenceCount = 0

    init(model: RPPGModel) { self.model = model }

    func reset() {
        ring.reset()
        tracker.reset()
        frameCounter = 0
        detectCounter = 0
        lastBox = nil
        setBusy(false)
        inferenceQueue.async {
            self.rawWaveform.removeAll()
            self.displayWaveform.removeAll()
            self.bpmFilter.reset()
        }
    }

    /// Called on the capture queue for every frame. The whole body runs inside an
    /// autoreleasepool: this queue is never idle, so without an explicit pool the
    /// autoreleased temporaries from Vision + Core Video would accumulate for the
    /// entire session and eventually get the app jetsam-killed.
    func process(_ pixelBuffer: CVPixelBuffer) {
        autoreleasepool {
            let bufferSize = CGSize(
                width: CVPixelBufferGetWidth(pixelBuffer),
                height: CVPixelBufferGetHeight(pixelBuffer)
            )

            // Detect + publish face state every 3rd frame (~10 Hz) for stability and CPU
            // headroom — but still crop/buffer EVERY frame so the model gets full 30 fps.
            detectCounter &+= 1
            let doDetect = detectCounter % 3 == 0
            if doDetect { lastBox = tracker.detect(pixelBuffer) }

            guard let box = lastBox else {
                // Face lost (beyond the tracker's hysteresis): drop the partial window so we
                // never stitch across a temporal gap. The HR reading persists briefly.
                ring.reset()
                frameCounter = 0
                if doDetect {
                    emitFace(FaceState(faceFound: false, faceBox: nil, bufferSize: bufferSize,
                                       lowLight: false, framesFilled: ring.filled, capacity: AppConfig.framesIn))
                }
                return
            }

            guard let cropped = FrameCropper.crop(pixelBuffer, roi: box) else { return }
            let lowLight = cropped.meanLuma < AppConfig.lowLumaThreshold
            ring.append(cropped.planarRGB)
            frameCounter &+= 1

            if doDetect {
                emitFace(FaceState(faceFound: true, faceBox: box, bufferSize: bufferSize,
                                   lowLight: lowLight, framesFilled: ring.filled, capacity: AppConfig.framesIn))
            }

            if ring.isFull, frameCounter >= AppConfig.stride, !isBusy() {
                frameCounter = 0
                setBusy(true)
                guard let snapshot = ring.snapshot() else { setBusy(false); return }
                inferenceQueue.async { [weak self] in self?.runInference(snapshot) }
            }
        }
    }

    // MARK: - Inference (inference queue)

    private func runInference(_ snapshot: [Float]) {
        defer { setBusy(false) }
        // Pool drains the large input/output tensors and the CoreML/Espresso runtime's
        // autoreleased intermediate buffers at the end of every window, so memory returns
        // to baseline between inferences instead of growing across the session.
        autoreleasepool {
            guard let tensor = TensorBuilder.build(snapshot) else { return }
            let chunkOut: [Float]
            do {
                chunkOut = try model.infer(tensor)   // `chunk` raw rPPG samples
            } catch {
                return   // transient failure: skip this window
            }

            inferenceCount += 1
            if inferenceCount % 10 == 1 {
                MemoryProbe.log("infer #\(inferenceCount)")
            }

            // Stitch this chunk's samples into the rolling analysis buffer.
            rawWaveform.append(contentsOf: chunkOut)
            if rawWaveform.count > AppConfig.analysisSamples {
                rawWaveform.removeFirst(rawWaveform.count - AppConfig.analysisSamples)
            }

            let progress = min(Double(rawWaveform.count) / Double(AppConfig.minAnalysisSamples), 1)

            var shownBPM = bpmFilter.value
            var quality = 0.0
            // Estimate over the stitched buffer once there's enough signal (~2 s).
            if rawWaveform.count >= 60, let result = HeartRateEstimator.estimate(rawWaveform, fs: AppConfig.fps) {
                displayWaveform = Array(result.filtered.suffix(AppConfig.analysisSamples))
                quality = result.quality
                // Update the shown HR on ANY physiological peak once warmed up — the
                // median+EMA filter handles noise, and `quality` drives the badge separately.
                // (Don't hard-gate the number on quality, or a weak-but-valid pulse shows nothing.)
                if rawWaveform.count >= AppConfig.minAnalysisSamples,
                   result.bpm >= AppConfig.minBPM, result.bpm <= AppConfig.maxBPM {
                    shownBPM = bpmFilter.update(result.bpm)
                }
            }

            let update = VitalsUpdate(bpm: shownBPM, quality: quality,
                                      waveform: displayWaveform, latencyMs: model.lastLatencyMs,
                                      warmupProgress: progress)
            DispatchQueue.main.async { [weak self] in
                guard let self else { return }
                self.delegate?.pipeline(self, didUpdateVitals: update)
            }
        }
    }

    // MARK: - Helpers

    private func emitFace(_ state: FaceState) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            self.delegate?.pipeline(self, didUpdateFace: state)
        }
    }

    private func isBusy() -> Bool { busyLock.lock(); defer { busyLock.unlock() }; return busy }
    private func setBusy(_ v: Bool) { busyLock.lock(); busy = v; busyLock.unlock() }
}

File: apps/Camera-Vitals/Android/app/src/main/java/ai/zetic/demo/cameravitals/pipeline/RppgPipeline.kt
package ai.zetic.demo.cameravitals.pipeline

import android.graphics.Bitmap
import android.graphics.Rect
import android.os.Handler
import android.os.Looper
import android.util.Size
import ai.zetic.demo.cameravitals.AppConfig
import ai.zetic.demo.cameravitals.signal.HeartRateEstimator
import ai.zetic.demo.cameravitals.util.MedianEMA
import ai.zetic.demo.cameravitals.util.MemoryProbe
import ai.zetic.demo.cameravitals.vision.FaceRoiTracker
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.math.min

class FaceState(
    val faceFound: Boolean,
    val faceBox: Rect?,
    val bufferSize: Size,
    val lowLight: Boolean,
    val framesFilled: Int,
    val capacity: Int
)

class VitalsUpdate(
    val bpm: Double?,
    val quality: Double,
    val waveform: FloatArray,
    val latencyMs: Double,
    val warmupProgress: Double
)

interface RppgPipelineListener {
    fun onFaceUpdate(state: FaceState)
    fun onVitalsUpdate(update: VitalsUpdate)
}

/**
 * Orchestrates the temporal pipeline: face crop → ring buffer → sliding-window inference → HR.
 * `process()` runs on the camera analyzer thread; inference runs on its own thread with a
 * single-flight `busy` gate (drop windows under back-pressure, never queue them).
 * Mirrors the iOS RPPGPipeline exactly.
 */
class RppgPipeline(private val model: RppgModel) {
    var listener: RppgPipelineListener? = null

    private val tracker = FaceRoiTracker()
    private val ring = FrameRingBuffer(AppConfig.FRAMES_IN, AppConfig.frameFloatCount)
    private val inferenceExecutor = Executors.newSingleThreadExecutor()
    private val main = Handler(Looper.getMainLooper())

    private var frameCounter = 0
    private var detectCounter = 0
    private var lastBox: Rect? = null
    private val busy = AtomicBoolean(false)

    // Stitched samples + display waveform + BPM smoothing — touched only on the inference thread.
    private var rawWaveform = FloatArray(0)
    private var displayWaveform = FloatArray(0)
    private val bpmFilter = MedianEMA(5, 0.3)
    private var inferenceCount = 0

    fun reset() {
        ring.reset()
        tracker.reset()
        frameCounter = 0
        detectCounter = 0
        lastBox = null
        busy.set(false)
        inferenceExecutor.execute {
            rawWaveform = FloatArray(0)
            displayWaveform = FloatArray(0)
            bpmFilter.reset()
        }
    }

    /** Called on the camera analyzer thread for every frame. Recycles `bitmap` before returning. */
    fun process(bitmap: Bitmap) {
        try {
            val bufferSize = Size(bitmap.width, bitmap.height)

            // Detect + publish face state every 3rd frame (~10 Hz); crop EVERY frame for 30 fps.
            detectCounter++
            val doDetect = detectCounter % 3 == 0
            if (doDetect) lastBox = tracker.detect(bitmap)

            val box = lastBox
            if (box == null) {
                // Face lost beyond hysteresis: drop the partial window so we never stitch a gap.
                ring.reset()
                frameCounter = 0
                if (doDetect) {
                    emitFace(FaceState(false, null, bufferSize, false, ring.filled, AppConfig.FRAMES_IN))
                }
                return
            }

            val cropped = FrameCropper.crop(bitmap, box) ?: return
            val lowLight = cropped.meanLuma < AppConfig.LOW_LUMA_THRESHOLD
            ring.append(cropped.planarRGB)
            frameCounter++

            if (doDetect) {
                emitFace(FaceState(true, box, bufferSize, lowLight, ring.filled, AppConfig.FRAMES_IN))
            }

            if (ring.isFull && frameCounter >= AppConfig.STRIDE && busy.compareAndSet(false, true)) {
                frameCounter = 0
                val snapshot = ring.snapshot()
                if (snapshot == null) {
                    busy.set(false)
                } else {
                    inferenceExecutor.execute { runInference(snapshot) }
                }
            }
        } finally {
            bitmap.recycle()
        }
    }

    private fun runInference(snapshot: FloatArray) {
        try {
            val tensor = TensorBuilder.build(snapshot) ?: return
            val chunkOut = try {
                model.infer(tensor)
            } catch (e: Exception) {
                return   // transient failure: skip this window
            }

            inferenceCount++
            if (inferenceCount % 10 == 1) MemoryProbe.log("infer #$inferenceCount")

            // Stitch this chunk's samples into the rolling analysis buffer.
            rawWaveform += chunkOut
            if (rawWaveform.size > AppConfig.ANALYSIS_SAMPLES) {
                rawWaveform = rawWaveform.copyOfRange(rawWaveform.size - AppConfig.ANALYSIS_SAMPLES, rawWaveform.size)
            }

            val progress = min(rawWaveform.size.toDouble() / AppConfig.MIN_ANALYSIS_SAMPLES, 1.0)

            var shownBPM = bpmFilter.value
            var quality = 0.0
            if (rawWaveform.size >= 60) {
                val result = HeartRateEstimator.estimate(rawWaveform, AppConfig.FPS)
                if (result != null) {
                    displayWaveform = result.filtered.takeLast(AppConfig.ANALYSIS_SAMPLES).toFloatArray()
                    quality = result.quality
                    // Update the shown HR on ANY physiological peak once warmed up — the
                    // median+EMA filter handles noise, quality drives the badge separately.
                    if (rawWaveform.size >= AppConfig.MIN_ANALYSIS_SAMPLES &&
                        result.bpm >= AppConfig.MIN_BPM && result.bpm <= AppConfig.MAX_BPM
                    ) {
                        shownBPM = bpmFilter.update(result.bpm)
                    }
                }
            }

            val update = VitalsUpdate(shownBPM, quality, displayWaveform, model.lastLatencyMs, progress)
            emitVitals(update)
        } finally {
            busy.set(false)
        }
    }

    private fun emitFace(state: FaceState) {
        main.post { listener?.onFaceUpdate(state) }
    }

    private fun emitVitals(update: VitalsUpdate) {
        main.post { listener?.onVitalsUpdate(update) }
    }
}
